### PR TITLE
Downstream koji build

### LINKED ISSUE #55: support building in koji for new dist-git commits
Support koji builds for new dist-git commits.

- [ ] Verify, that we accept new events for dist-git commits.
- [ ] Implement the handler (decide with the team):
    - [ ] Either enhance the KojiBuildHandler to differentiate between upstream and downstream git forge and act differently.
    - [ ] Or, create a new handler and new job type.

Edit: complete rewrite v2 by @lachmanfrantisek

### PR DESCRIPTION
**Pull request by lachmanfrantisek**
* Do not work with fedmsg topics in handlers.
    * The `PROCESSED_FEDMSG_TOPICS` wasn't used anywhere and @add_topic is not needed because of it.
* Implement the downstream Koji build.
    * Using a new `koji_build` job type.
* Make the `SyncFromDownstream` react to a general event:
    * Removing of `DistGitCommitEvent` that filtered the events based on the service config.
        * Events should be independent of handlers and multiple handlers can react to one event.
        * Logic (checking the service config) should be in a handler.
    * Removing FedmsgHandler because it had no function.

TODO:

- [x] feedback and error handling: Commit comment in case of an error on submission; once submitted, users will get a standard notification.
- [x] tests

Fixes #55

Merge after: packit/packit#1415 (required for tests to succeed)

When reviewing, please, go commit by commit. There were some class removals/renames that make the whole diff a bit messy.

---

You can set up a new `koji_build` job using the `commit` trigger to submit a Koji build for a new commit in a dist-git branch. The configuration file needs to be present in the dist-git for now (the state for the new commit is used).

### CODE REVIEW

File: packit_service/config.py [hunk 1]
```diff
@@ -251,6 +251,7 @@ def get_service_config(cls) -> "ServiceConfig":
         return cls.service_config
 
     def get_project_to_sync(self, dg_repo_name, dg_branch) -> Optional[ProjectToSync]:
+        # TODO: Is it ok that we don't check namespace? Can't this be misused from a fork?
```
Comment (TomasTomecek) [inline packit_service/config.py:254]: in such a case, the build should fail - one should not be able to submit production builds from a fork

File: packit_service/worker/events/event.py [hunk 2]
```diff
@@ -349,11 +349,14 @@ def get_package_config(self) -> Optional[PackageConfig]:
         )
 
         spec_path = None
-        if isinstance(self.base_project, PagureProject):
+        if self.project_url and RepoUrl.parse(self.project_url).hostname in [
+            "git.centos.org",
+            "git.stg.centos.org",
+        ]:
```
Comment (TomasTomecek) [inline packit_service/worker/events/event.py:355]: +1, this only applies to centos 7 and 8

File: packit_service/worker/handlers/distgit.py [hunk 3]
```diff
@@ -223,3 +242,74 @@ def run(self) -> TaskResults:
...
+            package_config=package_config,
+            job_config=job_config,
+            event=event,
+        )
+        self.dg_branch = event.get("git_ref")
+
+    def pre_check(self) -> bool:
+        if self.data.event_type in (PushPagureEvent.__name__,):
+            if self.data.git_ref != (
+                configured_branch := self.job_config.metadata.branch or "main"
```
Comment (TomasTomecek) [inline packit_service/worker/handlers/distgit.py:272]: WALRUS DETECTED!!

W

File: packit_service/config.py (hunk 1 above)
Comment (lachmanfrantisek) [reply packit_service/config.py:254]: But this is about upstream->downstream sync that I had to fix.

Comment (TomasTomecek) [reply packit_service/config.py:254]: ah, sorry, did not understand it then

**Comment by lachmanfrantisek** [comment]
recheck

**Comment by lachmanfrantisek** [comment]
recheck

**Comment by lachmanfrantisek** [comment]
recheck

**Review by TomasTomecek — APPROVED** [review]

**Pull request merged**